### PR TITLE
Externalize lazy execution semantic for EventExecutors

### PR DESCRIPTION
**Pull request by njhill**
Motivation

This is already done internally for various reasons but it would make sense i.m.o. as a top level concept: submitting a task to be run on the event loop which doesn't need to run immediately but must still be executed in FIFO order relative all other submitted tasks (be those "lazy" or otherwise).

It's nice to separate this abstract "relaxed" semantic from concrete implementations - the simplest is to just delegate to existing `execute`, but for the main EL impls translates to whether a wakeup is required after enqueuing.

Having a "global" abstraction also allows for simplification of our internal use - for example encapsulating more of the common scheduled future logic within `AbstractScheduledEventExecutor`.

Modifications

- Introduce public `LazyRunnable` interface and `AbstractEventExecutor#lazyExecute` method (would be nice for this to be added to `EventExecutor` interface in netty 5)
- Tweak existing `SingleThreadEventExecutor` mechanics to support these
- Replace internal use of `NonWakeupRunnable` (such as for pre-flush channel writes)
- Uplift scheduling-related hooks into `AbstractScheduledEventExecutor`, eliminating intermediate `executeScheduledRunnable` method

Result

Simpler code, cleaner and more useful/flexible abstractions - cleaner in that they fully communicate the intent in a more general way, without implying/exposing/restricting implementation details

### CODE REVIEW

**Comment by njhill** [comment]
This is just a proposal for further discussion, prompted by some comments exchanged with @Scottmitch here: https://github.com/netty/netty/pull/7834#discussion_r325822764. Although a little specialized, users/apps could also take advantage of it. It's kind of analagous to atomic lazySet :)

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill I wonder if this is a bit "too advanced". IMHO we should aim to keep the API surface low when there is no real demand for something.

**Comment by njhill** [comment]
@normanmaurer that's fair. But this is already used from a different package within netty, so even making it `protected` means having the package-private subinterface of `NonWakeupRunnable` that's currently in `SingleThreadEventLoop`, which is kind of hacky imo. It might be that there's other places in the code where `execute`s can be changed to `lazyExecute`s (or `LazyRunnable`s used in place of `Runnable`s)... I will see if I can spot any.

Even if we don't add any public interfaces, WDYT about keeping it at the `AbstractEventExecutor` level i.e. keeping the other changes in this PR but changing those be `protected` + using aforementioned subinterface workaround?

**Comment by njhill** [comment]
@normanmaurer I would like to lobby a bit more for this one. It eliminates an allocation on the future scheduling/cancellation paths, and I have an additional optimization based on top of it here https://github.com/njhill/netty/pull/1 which eliminates some further allocations.

Combined with #9666 this means quite a bit less garbage when scheduling/cancelling tasks, and gives noticeable performance improvement (I think this is mostly independent of GC reduction):

Before:
```
Benchmark                       (num)   Mode  Cnt    Score   Error  Units
scheduleCancelLotsOutsideLoop  100000  thrpt   10  131.840 ± 2.643  ops/s
scheduleLotsOutsideLoop        100000  thrpt   10  233.181 ± 9.407  ops/s
```

After:
```
Benchmark                       (num)   Mode  Cnt    Score    Error  Units
scheduleCancelLotsOutsideLoop  100000  thrpt   10  154.743 ±  0.953  ops/s
scheduleLotsOutsideLoop        100000  thrpt   10  278.944 ± 12.727  ops/s
```

It might also be useful for implementing #7719.

**Comment by normanmaurer** [comment]
@njhill yeah keeping it protected for now sounds fine to me...

**Comment by normanmaurer** [comment]
@njhill please ping me once done

**Comment by njhill** [comment]
@normanmaurer I changed the new method and interface to be protected and moved the interface from within `EventExecutor` to `AbstractEventExecutor`. Also added test coverage.

But, I still think it would make much more sense for these to be public... the main use of such a method (`lazyExecute`) is to be used from outside an event loop by arbitrary code that makes use of that event loop. Having it protected basically precludes most of such usage, and as mentioned means reintroducing a hack where an EventExecutor subclass in transport package needs to define a separate subinterface to provide visibility to other classes in the transport package.

It's not _really_ extending the API surface since the API is defined by `EventExecutor` and that isn't changed (yet). I know it still means we can't then remove or break it in future but IMHO that's not much of a concern here.

**Comment by johnou** [comment]
Public with unstableapi annotation?

**Comment by njhill** [comment]
@johnou sure that's also an option. @normanmaurer WDYT?

**Comment by normanmaurer** [comment]
@njhill if you are strong about it sure...

**Comment by johnou** [comment]
@njhill is there a threshold of max queued lazy tasks before a wake is triggered?

**Comment by njhill** [comment]
@johnou there is not a max threshold but imho it shouldn't be a problem as this is only for use in specific circumstances and _any_ subsequent event loop activity will cause them all to be flushed.

@normanmaurer now updated back to public within `AbstractEventExecutor`, marked with `@UnstableApi` as suggested by @johnou.

After this is in I will open https://github.com/njhill/netty/pull/1 against 4.1 branch.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@johnou PTAL

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java [hunk 1]
```diff
@@ -165,4 +166,23 @@ protected static void safeExecute(Runnable task) {
...
+    /**
+     * Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
+     * a non-lazy task is executed or the executor is shut down.
+     *
+     * This is equivalent to submitting a {@link EventExecutor.LazyRunnable} to
+     * {@link #execute(Runnable)} but for an arbitrary {@link Runnable}.
+     */
+    @UnstableApi
+    public void lazyExecute(Runnable task) {
+        execute(task);
```
Comment (johnou) [inline common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: shouldn't it be wrapped in LazyRunnable..?

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java [hunk 2]
```diff
@@ -243,12 +248,22 @@ protected void validateScheduled(long amount, TimeUnit unit) {
...
                 @Override
                 public void run() {
                     scheduledTaskQueue().add(task.setId(nextTaskId++));
                 }
-            }, true, task.deadlineNanos());
+            };
+            if (beforeScheduledTaskSubmitted(deadlineNanos)) {
+                execute(addToQueue);
+            } else {
+                lazyExecute(addToQueue);
```
Comment (johnou) [inline common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: in theory, if I constantly write from outside the event loop fast enough the mpsc queue will blow up, right?

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: I don't think that should be necessary - since `LazyRunnable` is a new interface, if some subclass was to support it then it could/would also support `lazyExecute` and there would be no need for the extra wrapping (the motivation for having a separate method in the first place).

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: The usage here is equivalent to the current behaviour, just avoids the extra wrapper Runnable. 

I gave my thoughts w.r.t general use of `lazyExecute` in response to your earlier question. Of course the method could be used to break things but there are lots of ways to break things. For example if you schedule a gazillion future tasks it will blow up the pending task queue (ignoring delayed execution). Or if you schedule tasks onto the event loop from multiple other threads faster than it can keep up with them, etc.

**Comment by normanmaurer** [comment]
@johnou PTAL again

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: @johnou is this resolved ?

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: @johnou is this resolved ?

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: hard to follow the code flow in GitHub, my main point is there is a method called lazyExecute that takes Runnable, which invokes execute directly.. I would have thought we'd need something like this.. 

```java
    @UnstableApi
    public void lazyExecute(final Runnable task) {
        if (task instanceof LazyRunnable) {
            execute(task);
        } else {
            execute(new LazyRunnable() {
                @Override
                public void run() {
                    task.run();
                }
            });
        }
    }
```

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: Well it's just a theory and I haven't tested it, also invoking size() on the mpsc to track pending tasks is probably expensive. I am happy to test it out with our application to see how it behaves.

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: @njhill ptal

Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: @johnou it is not an implementation per se but rather just default behaviour for (possibly pre-existing) `AbstractEventExecutor` subclasses which don't support this lazy execution semantic. I.e. like `default` interface method in java8+. The method itself is just for providing a "hint" to allow for internal optimization (delayed wakeup) ... there should never be any problem _functionally_ with ignoring the lazy aspect and running immediately.

The idea is to provide some flexiblity in how the laziness is communicated - it can be via use of this method _and/or_ by implementing `LazyRunnable`. The former mechanism allows for deciding this for arbitrary tasks at the time of submission without having to wrap in a separate Runnable. Supporting implementations would recognize both (see the changes to `SingleThreadEventExecutor`).

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: > invoking size() on the mpsc to track pending tasks is probably expensive

@johnou could you elaborate on this? I'm not sure how it applies to the changes in this PR. Also I don't think I would call it expensive if you are thinking of the JCTools impl that we use, especially when called from the consumer thread.

Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: I must be mixing it up with https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedDeque.html#size() where it's not a constant time operation.

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: Ok so the default implementation breaks the API contract to maintain backwards compatibility.

Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: No API contracts are being broken :) The regular `execute` method implies "stronger" guarantees that `lazyExecute` does (i.e. that the execution will be scheduled immediately), and thus there is no problem defaulting to existing implementations of that... it's a perfectly valid way to perform `lazyExecute`. It just doesn't happen to use the extra slack afforded to it by the "lazy" part.

Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: I was more concerned about the javadoc, would you be against updating the wording from "Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
     * a non-lazy task is executed or the executor is shut down.
     *" to something like.. "Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
     * a non-lazy task is executed or the executor is shut down. If the implementing AbstractEventExecutor does not support lazy tasks it will fallback to the default execute behaviour.
     *"

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: @johnou so can this be resolved ?

File: common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java]: @njhill ^^?

File: common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java (hunk 2 above)
Comment (johnou) [reply common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java]: Yes I'm happy with the explanation and fact it is marked unstableapi.

File: transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java [hunk 3]
```diff
@@ -159,9 +159,4 @@ public int pendingTasks() {
     public int registeredChannels() {
         return -1;
     }
-
-    /**
-     * Marker interface for {@link Runnable} that will not trigger an {@link #wakeup(boolean)} in all cases.
-     */
-    interface NonWakeupRunnable extends SingleThreadEventExecutor.NonWakeupRunnable { }
```
Comment (johnou) [inline transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java]: Doesn't this need to stay / mark deprecated?

**Comment by normanmaurer** [comment]
@netty-bot test this please
